python/ray/train/v2/_internal/execution/checkpoint/validation_manager.py

```diff
@@ -0,0 +1,108 @@
+import logging
+import time
+from collections import OrderedDict
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+import ray
+from ray.train._checkpoint import Checkpoint
+from ray.train.v2._internal.execution.callback import ControllerCallback, ReportCallback
+from ray.train.v2._internal.execution.checkpoint.checkpoint_manager import (
+    CheckpointManager,
+)
+from ray.train.v2._internal.execution.training_report import _ValidationSpec
+from ray.train.v2.api.exceptions import ValidationFailedError
+from ray.train.v2.api.validation_info import ValidationInfo
+
+if TYPE_CHECKING:
+    from ray.train.v2._internal.execution.controller import TrainControllerState
+
+logger = logging.getLogger(__name__)
+
+
+VALIDATION_TASK_POLL_INTERVAL_S = 1
+
+
+@ray.remote
+def run_validate_fn(validation_spec: _ValidationSpec, checkpoint: Checkpoint) -> Dict:
+    """Run the user-defined validation function."""
+    metrics_dict = validation_spec.validate_fn(
+        checkpoint,
+        validation_spec.validate_config,
+    )
+    if not isinstance(metrics_dict, dict):
+        raise ValueError(
+            "The validate function must return a dictionary of metrics. "
+            f"Got {type(metrics_dict)} instead."
+        )
+    return metrics_dict
+
+
+class ValidationManager(ControllerCallback, ReportCallback):
+    def __init__(
+        self,
+        checkpoint_manager: CheckpointManager,
+    ):
+        self._checkpoint_manager = checkpoint_manager
+
+        # Map from checkpoint to validation task
+        self._pending_validations = {}
+
+        # Map from checkpoint to ValidationInfo
+        self._failed_validations = OrderedDict()
```

Comment:
add todo for checkpointing validation manager state

Reply:
done.